File: DataStructures/Search&Sort/Sorting/BISHSort.cpp
/*Insertion Sort*/
#include<iostream>
using namespace std;
//Function to swap 2 numbers
void swap(int * a, int* b){
	int temp;
	temp=*a;
	*a=*b;
	*b=temp;
}
//Insertion Sort Function
void insertionSort(int a[],int n){
	int i,j,min;
	for(i=1;i<n;i++){
		min=a[i];
		j=i-1;
		while(j>=0&&a[j]>min){
			a[j+1]=a[j];
			j--;
		}
		a[j+1]=min;
	}
}

//Bubble Sort
void bubble(int arr[], int n){
	for(int i=0;i<n;i++){
		cout<<"i "<<i<<endl;
		for(int j=0;j<n-i-1;j++){
			cout<<"\tj "<<j<<endl;
			if(arr[j]>arr[j+1]){
				swap(&arr[j],&arr[j+1]);
			}
		}
	}
	printarr(arr,n);
}

//Selection Sort
void selectionSort(int arr[],int n){
	int min;
	for(int i=0;i<n-1;i++){
		min=i;
		for(int j=i+1;j<n;j++)
			if(arr[j]<arr[min])
				min=j;
		swap(&arr[min],&arr[i]);
	}
	printarr(arr,n);
}

//Function to Keep the Tree Heapify or matching the condition
void heapify(int arr[],int n, int i){
	int largest=i;
	int l=2*i+1;
	int r=2*i+2;
	if(l<n && arr[l]>arr[largest]){
		largest=l;
	}
	if(r<n && arr[r]>arr[largest]){
		largest=r;
	}
	if(largest!=i){
		swap(&arr[i],&arr[largest]);
		heapify(arr,n,largest);
	}
}

//Function to sort the Heap
void heapsort(int arr[], int n){
	for(int i=n/2-1;i>=0;i--){
		heapify(arr,n,i);
	}
	for(int j=n-1;j>=0;j--){
		swap(&arr[0],&arr[j]);
		heapify(arr,j,0);
	}
}
//Prints an array
void printarr(int a[],int n){
	for(int l=0;l<n;l++){
		cout<<"\t"<<a[l];
	}
}
int main(){
	int a[]={10,9,8,1,6,5,4};
	int size=((sizeof(a))/sizeof(a[0]));
	insertionSort(a,size);
	printarr(a,size);
}

File: DataStructures/Search&Sort/Sorting/BISHSort.py
import time
def printarray(array,size):
    for i in range(size):
        print(array[i],end=" ")

def bubblesort(array,size):
    for i in range(size):
        for j in range(size-1):
            if array[j]>array[j+1]:
                array[j],array[j+1]=array[j+1],array[j]
    printarray(array,size)
    
def selectionsort(array,size):
    for i in range(size):
        min=i
        for j in range(size):
            if array[min]>array[j]:
                min=j
        array[i],array[min]=array[min],array[i]
    printarray(array,size)

def insertionsort(array,size):
    for i in range(1,size):
        key=array[i]
        j=i-1
        while j>=0 and key<array[j]:
            array[j+1]=array[j]
            j-=1
        array[j+1]=key
    printarray(array,size)

def heapify(array,size,i):
    largest=i
    left=2*i+1
    right=2*i+2

    if left<size and array[left]>array[i]:
        largest=left
    if right<size and array[right]>array[largest]:
        largest=right
    if largest!=i:
        array[i],array[largest]=array[largest],array[i]
        heapify(array,size,largest)

def heapSort(array,size):
    for i in range(size,-1,-1):
        heapify(array,size,i)
    for i in range(size-1,0,-1):
        array[0],array[i]=array[i],array[0]
        heapify(array,i,0)
    return array

def main():
    array=[98,12,45,67,13,43,77]
    size=len(array)
    
    t1=time.time()
    bubblesort(array,size)
    t2=time.time()
    print("\nTime taken for Bubble Sort {0}".format(round(t2-t1,4)))
    
    t1=time.time()
    selectionsort(array,size)
    t2=time.time()
    print("\nTime taken for Selection Sort {0}".format(round(t2-t1,4)))

    t1=time.time()
    insertionsort(array,size)
    t2=time.time()
    print("\nTime taken for Insertion Sort {0}".format(round(t2-t1,4)))
    

main()